src/firebolt_db/firebolt_async_dialect.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+from types import ModuleType
+from typing import Any, Iterator, List, Optional, Tuple
+
+import firebolt.async_db as async_dbapi
+from firebolt.async_db import Connection
+from sqlalchemy.engine import AdaptedConnection  # type: ignore[attr-defined]
+from sqlalchemy.util.concurrency import await_only
+
+from firebolt_db.firebolt_dialect import FireboltDialect
+
+
+class AsyncCursorWrapper:
+    __slots__ = (
+        "_adapt_connection",
+        "_connection",
+        "description",
+        "await_",
+        "_cursor",
+        "_rows",
+        "arraysize",
+        "rowcount",
+    )
+
+    server_side = False
+
+    def __init__(self, adapt_connection: AsyncConnectionWrapper):
+        self._adapt_connection = adapt_connection
+        self._connection = adapt_connection._connection
+        self.await_ = adapt_connection.await_
+        self.arraysize = 1
+        self.rowcount = -1
+        self.description = None
+        self._rows: List[Tuple] = []
+
+    def close(self) -> None:
+        self._rows[:] = []
+
+    def execute(self, operation: str, parameters: Optional[Tuple] = None) -> None:
+        _cursor = self._connection.cursor()
```

Comment:
Shouldn't we create a single cursor in `__init__()` and that close it in `close()`?
I guess this won't make any functional changes now but it seems like that's how it should wrap

Reply:
Also, this way we won't need to store `description` and `rowcount` but just simply return `_cursor.description` and `_cursor.rowcount`

Reply:
True, so far I tested all or nothing approach by replacing everything with our cursor and instantiating it in the __init__. However, that had issues with `fetchall`, etc. being called from a sync context. Potentially, a hybrid can work, where we store `_rows` but everything else uses the shared cursor. Need to be careful with the thread safety though, not sure if sqlalchemy shares the cursor across.

Reply:
Well, I don't have a strong opinion here but this approach just seems more natural to me. Thread safety shouldn't be an issue but I can't be sure since there are still no tests for it. It's up to you though

Reply:
I still think thread safety can be an issue so using postgres and mysql dialects as an example I've added a mutex for execution. Apart from that yeah, it looks better now that I changed the wrapper to reuse the cursor.